### PR TITLE
Blog: Using MQTT Sparkplug B with Node-RED

### PR DESCRIPTION
**Pull request by sumitshinde-84**
## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] I have considered the performance impact of these changes
 - [ ] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated

### CODE REVIEW

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+date: 2024-08-21
+authors: ["sumit-shinde"]
+image: 
+tags:
+   - post
+   - nodered
+   - mqtt sparkplug b with node-red
+---
+
+Connected devices are creating a lot of data, but without a standard format, managing this data can be tricky. MQTT helps with messaging, but it can be inconsistent. MQTT Sparkplug B solves this problem by providing a clear, standardized format for data. In this guide, we’ll show you how to use MQTT Sparkplug B with Node-RED to make managing your device data easier and more organized.
```
Comment (ZJvandeWeg) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:14]:
```suggestion
Connected devices are creating a lot of data but without a standard format, managing this data can be tricky. MQTT helps with messaging, though MQTT is flexible and doesn't standardize message formats. MQTT Sparkplug B solves this problem by providing a clear, standardized format for data. In this guide, we’ll show you how to use MQTT Sparkplug B with Node-RED to make managing your device data easier and more organized.
```

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+<!--more-->
+
+## What is MQTT Sparkplug?
+
+*Sparkplug B* is an open-source specification hosted at the Eclipse Foundation that extends MQTT by providing a standardized way to structure data. It ensures that all devices use a consistent message format, that simplifies integration and improves data consistency.
```
Comment (ZJvandeWeg) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:20]: I think a key point is that it's optimized in terms of structure for sensor readings and device data being published. It's not a standard for pushing changes to a device or sensor.

**Comment by ZJvandeWeg** [comment]
@Steve-Mcl I think you have more experience than I have with Sparkplug B, can you review? If not, that's fine, and I'll research a bit

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+- **[Group ID]**: Represents the group or application context.
+- **[Message Type]**: Specifies the type of message (e.g., NBIRTH, NDATA).
+- **[EON Node ID]**: Identifies the edge node or system.
+- **[Device ID]**: Identifies the specific device (if applicable).
+
+## Using MQTT Sparkplug with Node-RED
+
+In this section, we'll explore an example where Machine1 on the factory floor sends temperature and humidity data to an MQTT broker. We’ll cover how to integrate MQTT Sparkplug B into Node-RED and demonstrate basic operations such as sending and receiving data using MQTT Sparkplug B and more.
+
+### Prequsite
```
Comment (Steve-Mcl) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:82]:
```suggestion
### Prerequisite
```

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+Before you begin, ensure you have the following:
+
+- [node-red-contrib-mqtt-sparkplug-plus](https://flows.nodered.org/node/node-red-contrib-mqtt-sparkplug-plus): Install this Node-RED package for Sparkplug B support via pallete manager.
```
Comment (Steve-Mcl) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:86]:
```suggestion
- [node-red-contrib-mqtt-sparkplug-plus](https://flows.nodered.org/node/node-red-contrib-mqtt-sparkplug-plus): Install this Node-RED package for Sparkplug B support via palette manager.
```

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+2. Drag the mqtt sparkplug device node onto the canvas.
+3. Double-click the mqtt sparkplug device node to open the configuration panel. Add the metric names that you will be sending by clicking the bottom-left "Add" button. Ensure that the names match the metric names in the payload you are sending and specify the data types for each metric.
+
+![Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics](./images/mqtt-sparkplug-device-node.png "Screenshot showing the Sparkplug Device node configuration and the 'Add' button for defining metrics")  
+_Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics_
+
+4. Switch to the Advanced tab by clicking the "Advanced" option at the top-right.
+
+5. Enable the "Send Birth Immediately" option. This ensures that a Birth message (DBIRTH) is sent immediately upon deployment and connection to the MQTT broker. Note that enabling this option will send the `DBIRTH` message when the device node connects, but an `NBIRTH` message will be sent successfull connection of mqtt sparkplug out node if you are using.
```
Comment (Steve-Mcl) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:128]:
```suggestion
5. Enable the "Send Birth Immediately" option. This ensures that a Birth message (DBIRTH) is sent immediately upon deployment and connection to the MQTT broker. Note that enabling this option will send the `DBIRTH` message when the device node connects, but an `NBIRTH` message will be sent successful connection of mqtt sparkplug out node if you are using.
```

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+![Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics](./images/mqtt-spark-device-advance.png "Screenshot showing the Sparkplug Device node configuration and the 'Add' button for defining metrics")  
+_Screenshot showing the Sparkplug Device node configuration and the "Add" button for defining metrics_
+
+6. Optionally, enable Store and Forward when not connected to ensure that messages are stored and sent once the connection is re-established. To use this option, make sure you have enabled it in the mqtt sparkplug broker config node and specified the destination.
+7. Connect the inject node's output to the mqtt sparkplug device node's input.
+8. Deploy the flow by clicking the top-right "Deploy" button.
+
+Once you deploy the flow and all devices connect to the MQTT broker, the system automatically send a `DBIRTH` message as soon as each device within the node connects, signaling that the device is ready for data transmission.
```
Comment (Steve-Mcl) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:137]:
```suggestion
Once you deploy the flow and all devices connect to the MQTT broker, the system automatically send a `DBIRTH` message as soon as each device within the node connects, signalling that the device is ready for data transmission.
```

**Review by Steve-Mcl — CHANGES_REQUESTED** [review]
A few typos and suggestions.

Please do not accept my suggestions if they do not make sense (I am not fully up to speed on Sparkplug.


NOTE: There are about 20+ places there `sparkplug` should be written `Sparkplug` - the spec always refers to `Sparkplug` with a capital "S" - please consider updating the rest of the doc accordingly.

PS, great doc 👍

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md (hunk 1 above)
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:14]: NB: A counter suggestion was offered in my review here: https://github.com/FlowFuse/website/pull/2478#discussion_r1724909709

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md (hunk 2 above)
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:20]: NB: A counter suggestion was offered in my review here: https://github.com/FlowFuse/website/pull/2478#discussion_r1724915459

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:20]: "improves data consistency" didnt "feel right"

In terms of data, consistency is often related to timing and factor not necessarily related to "structure".

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:20]: This re-write removes the "simplifies things" and sticks to the official line from ecplise.  the "simplifies things" is touched on in the next sentence so this was somewhat superfluous IMHO.

**Comment by sumitshinde-84** [comment]
> accordingly



> A few typos and suggestions.
> 
> Please do not accept my suggestions if they do not make sense (I am not fully up to speed on Sparkplug.
> 
> NOTE: There are about 20+ places there `sparkplug` should be written `Sparkplug` - the spec always refers to `Sparkplug` with a capital "S" - please consider updating the rest of the doc accordingly.
> 
> PS, great doc 👍

Thanks a lot, @Steve-Mcl, for the review. You made a huge improvement! I’ll also make sure to update `sparkplug` as you suggested.

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md (hunk 1 above)
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: outdated

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md (hunk 2 above)
Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: outdated

**Review by Steve-Mcl — APPROVED** [review]
Looks good to me, but please do not merge until @ZJvandeWeg has also reviewed.

**Comment by sumitshinde-84** [comment]
Yes, we also need to update the images and push the tile before merging. I have created an art request for it: https://github.com/FlowFuse/website/issues/2487.

File: src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
+---
+title: Using MQTT Sparkplug B with Node-RED
+subtitle: A Practical Guide to integrating MQTT Sparkplug B with Node-RED
+description: Explore how MQTT Sparkplug B enhances MQTT with standardized data formats, making integration easier. Learn how to set up Sparkplug B in Node-RED for efficient data management in industrial environments.
+date: 2024-08-21
+authors: ["sumit-shinde"]
+image:
```
Comment (ZJvandeWeg) [inline src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md:7]: @sumitshinde-84 Please make a tile

Comment (sumitshinde-84) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: Done !

Comment (sumitshinde-84) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: we need to replace the diagrams !

Comment (Steve-Mcl) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: > we need to replace the diagrams !

Are you saying the article wasnt 100% ready @sumitshinde-84?

FWIW, i typically switch a PR to draft if not ready to go.

Comment (sumitshinde-84) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: The article is ready; it just needs the diagram updated with the diagrams Yndira will recreate.

Comment (sumitshinde-84) [reply src/blog/2024/08/using-mqtt-sparkplugb-with-node-red.md]: I created those rough diagrams so that Yndira can understand the type of diagram we need.

**Pull request merged**